spec.bs

```diff
@@ -223,6 +227,48 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
   </dd>
 </dl>
 
+<div algorithm>
+  The <dfn constructor for=URLPattern lt="URLPattern(input, baseURL)">new URLPattern(|input|, |baseURL|)</dfn> constructor steps are:
+
+  1. Let |init| be null.
+  1. If |input| is a string then:
+    1. Set |init| to the result of running [=parse a constructor string=] given |input|.
+    1. Set |init|["{{URLPatternInit/baseURL}}"] to |baseURL|.
+  1. Else:
+    1. [=Assert=]: |input| is a {{URLPatternInit}}.
+    1. If |baseURL| is given, then throw a {{TypeError}}.
+    1. Set |init| to |input|.
+  1. Let |protocol| be null.
```

Comment:
But the passed in values are significant; empty string vs null.  Also, this follows the pattern already in the match algorithm.

Reply:
Well, right now it's never null, because you always overwrite it on line 250.

Reply:
These values are passed as arguments to [=process a URLPatternInit=].

Reply:
Oh yeah... that makes no sense to me then, since you're always passing null. That algorithm doesn't even accept null; it requires those arguments to be strings.

Reply:
I think that algorithm has been changed to accept string or null now.  As we talked about offline, though, I will plan to do another pass to try to make "process a URLPatternInit" cleaner.  Its a bit awkward at the moment due to the translation from c++ where these arguments are passed by reference and modified in place.

Reply:
Even before that refactoring, I think this would be clearer if you got rid of lines 240-247, changed line 248 to just say "null, null, null, null, null, null, null, and null", and then used "Let .. be" instead of "Set ... to" on lines 249-256.

It's just really confusing right now to reuse the variables as both inputs and outputs.